File: Array-String/CPP/find_common_characters.cpp
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    vector<string> commonChars(vector<string>& words) {
        unordered_map<char, int> temp = counter(words[0]);
        for (int i = 1; i < words.size(); i++) {
            unordered_map<char, int> cur = counter(words[i]);
            unordered_map<char, int> calc;
            for (char w : intersect(cur, temp)) calc[w] = min(cur[w], temp[w]);
            temp = calc;
        }
        vector<string> ans;
        for (auto k : temp) {
            for (int i = 0; i < k.second; i++) ans.push_back(string {k.first});
        }
        return ans;
    }

    unordered_map<char, int> counter(string& word) {
        unordered_map<char, int> map;
        for (char w : word) map[w] += 1;
        return map;
    }

    vector<char> intersect(unordered_map<char, int>& map1, unordered_map<char, int>& map2) {
        vector<char> res;
        for (auto k: map1) {
            if (map2.find(k.first) != map2.end()) res.push_back(k.first);
        }
        return res;
    }
};

File: Array-String/Python/find_common_characters.py
from collections import Counter
from functools import reduce
from operator import and_

class Solution:
    def commonChars(self, words: list[str]) -> list[str]:
        temp = Counter(words[0])
        for i in range(1, len(words)):
            cur = Counter(words[i])
            calc = {w: min(cur[w], temp[w]) for w in (cur.keys() & temp.keys())}
            temp = calc
        ans = []
        for k in temp:
            ans.extend([k] * temp[k])
        return ans

    # shorter
    def commonChars(self, A: list[str]) -> list[str]:
        res = Counter(A[0])
        for a in A:
            res &= Counter(a)
        return list(res.elements())

    # one-liner
    def commonChars(self, A: list[str]) -> list[str]:
        return reduce(and_, map(Counter, A)).elements()